roles/pulp_common/templates/requirements.in.j2

```diff
@@ -1,6 +1,8 @@
 pulpcore=={{ __pulp_version }}
 {% for plugin, value in pulp_install_plugins_normalized.items() %}
-{% if value['source_dir'] is defined -%}
+{% if value['source_dir'] is defined and value['git_url'] is defined-%}
+git+{{ value['git_url']}}@{{ value['git_revision'] | default('master') }}#egg={{ plugin }}
+{% elif value['source_dir'] is defined -%}
```

Comment:
We still have this line though:https://github.com/pulp/pulp_installer/blob/master/roles/pulp_common/tasks/install_pip.yml#L124

So the only circumstance that this would be used is when:
1. pulp is installed from PyPI.
2. Plugins are installed from git.

Can it properly function when pulp is installed from git? Does pip-compile even understand the git URLs when it checks?

Reply:
Yes, you are correct about your asessment, this would be te only circumstance when that would be run.

Yes, pip-compile works as expected, below the proof.

```
[root@ip-10-0-15-31 pulp]# pwd
/usr/local/lib/pulp

[root@ip-10-0-15-31 pulp]# cat requirements.in 
pulpcore==3.14.2
pulp-container
pulp-ansible
git+https://github.com/ansible/galaxy_ng@master#egg=galaxy-ng

# Any plugins listed below were already installed but not specified in
# pulp_install_plugins

[root@ip-10-0-15-31 pulp]# bin/pip-compile 
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile
#
aiodns==3.0.0
    # via pulpcore
aiofiles==0.7.0
    # via pulpcore
aiohttp==3.7.4.post0
    # via pulpcore
aioredis==2.0.0
    # via pulpcore
ansible-core==2.11.3
    # via galaxy-importer
ansible-lint==5.1.2
    # via galaxy-importer
async-lru==1.0.2
    # via pulp-ansible
async-timeout==3.0.1
    # via
    #   aiohttp
    #   aioredis
asyncio-throttle==1.0.2
    # via pulpcore
attrs==21.2.0
    # via
    #   aiohttp
    #   galaxy-importer
    #   jsonschema
backoff==1.10.0
    # via pulpcore
bleach==3.3.1
    # via galaxy-importer
bleach-allowlist==1.0.3
    # via galaxy-importer
bracex==2.1.1
    # via wcmatch
certifi==2021.5.30
    # via requests
cffi==1.14.6
    # via
    #   cryptography
    #   pycares
chardet==4.0.0
    # via aiohttp
charset-normalizer==2.0.4
    # via requests
click==8.0.1
    # via
    #   pulpcore
    #   rq
colorama==0.4.4
    # via rich
commonmark==0.9.1
    # via rich
cryptography==3.4.7
    # via
    #   ansible-core
    #   pyjwt
defusedxml==0.7.1
    # via odfpy
diff-match-patch==20200713
    # via django-import-export
django==2.2.24
    # via
    #   django-currentuser
    #   django-filter
    #   django-guardian
    #   django-guid
    #   django-import-export
    #   django-lifecycle
    #   djangorestframework
    #   drf-nested-routers
    #   drf-spectacular
    #   galaxy-ng
    #   pulpcore
django-currentuser==0.5.3
    # via pulpcore
django-filter==2.4.0
    # via pulpcore
django-guardian==2.4.0
    # via pulpcore
django-guid==2.2.1
    # via pulpcore
django-import-export==2.5.0
    # via pulpcore
django-lifecycle==0.9.1
    # via pulpcore
django-prometheus==2.1.0
    # via galaxy-ng
djangorestframework==3.12.4
    # via
    #   drf-nested-routers
    #   drf-spectacular
    #   pulpcore
djangorestframework-queryfields==1.0.0
    # via pulpcore
drf-access-policy==0.9.2
    # via pulpcore
drf-nested-routers==0.93.3
    # via pulpcore
drf-spectacular==0.17.2
    # via
    #   galaxy-ng
    #   pulpcore
dynaconf==3.1.4
    # via pulpcore
ecdsa==0.13.3
    # via pulp-container
enrich==1.2.6
    # via ansible-lint
et-xmlfile==1.1.0
    # via openpyxl
flake8==3.9.2
    # via galaxy-importer
future==0.18.2
    # via pyjwkest
galaxy-importer==0.3.4
    # via
    #   galaxy-ng
    #   pulp-ansible
git+https://github.com/ansible/galaxy_ng@master#egg=galaxy-ng
    # via -r requirements.in
gunicorn==20.1.0
    # via pulpcore
idna==3.2
    # via
    #   requests
    #   yarl
inflection==0.5.1
    # via drf-spectacular
jinja2==3.0.1
    # via
    #   ansible-core
    #   pulpcore
jsonschema==3.2.0
    # via
    #   drf-spectacular
    #   pulp-ansible
markdown==3.3.4
    # via galaxy-importer
markuppy==1.14
    # via tablib
markupsafe==2.0.1
    # via jinja2
mccabe==0.6.1
    # via flake8
multidict==5.1.0
    # via
    #   aiohttp
    #   yarl
odfpy==1.4.1
    # via tablib
openpyxl==3.0.7
    # via tablib
packaging==21.0
    # via
    #   ansible-core
    #   ansible-lint
    #   bleach
    #   pulp-ansible
prometheus-client==0.11.0
    # via django-prometheus
psycopg2==2.8.6
    # via pulpcore
pulp-ansible==0.9.0
    # via
    #   -r requirements.in
    #   galaxy-ng
pulp-container==2.7.1
    # via
    #   -r requirements.in
    #   galaxy-ng
pulpcore==3.14.2
    # via
    #   -r requirements.in
    #   galaxy-ng
    #   pulp-ansible
    #   pulp-container
pycares==4.0.0
    # via aiodns
pycodestyle==2.7.0
    # via flake8
pycparser==2.20
    # via cffi
pycryptodomex==3.10.1
    # via pyjwkest
pyflakes==2.3.1
    # via flake8
pygments==2.9.0
    # via rich
pygtrie==2.4.2
    # via pulpcore
pyjwkest==1.4.2
    # via pulp-container
pyjwt[crypto]==1.7.1
    # via pulp-container
pyparsing==2.4.7
    # via
    #   drf-access-policy
    #   packaging
pyrsistent==0.18.0
    # via jsonschema
python-gnupg==0.4.7
    # via pulpcore
pytz==2021.1
    # via django
pyyaml==5.4.1
    # via
    #   ansible-core
    #   ansible-lint
    #   drf-spectacular
    #   galaxy-importer
    #   pulp-ansible
    #   pulpcore
    #   tablib
redis==3.5.3
    # via
    #   pulpcore
    #   rq
requests==2.26.0
    # via
    #   galaxy-importer
    #   pyjwkest
resolvelib==0.5.4
    # via ansible-core
rich==10.6.0
    # via
    #   ansible-lint
    #   enrich
rq==1.8.1
    # via pulpcore
ruamel.yaml==0.17.10
    # via ansible-lint
ruamel.yaml.clib==0.2.6
    # via ruamel.yaml
semantic-version==2.8.5
    # via
    #   galaxy-importer
    #   pulp-ansible
six==1.16.0
    # via
    #   bleach
    #   jsonschema
    #   pyjwkest
    #   url-normalize
sqlparse==0.4.1
    # via django
tablib[html,ods,xls,xlsx,yaml]==3.0.0
    # via django-import-export
tenacity==8.0.1
    # via ansible-lint
typing-extensions==3.10.0.0
    # via
    #   aiohttp
    #   aioredis
uritemplate==3.0.1
    # via drf-spectacular
url-normalize==1.4.3
    # via pulp-container
urllib3==1.26.6
    # via requests
urlman==2.0.1
    # via django-lifecycle
wcmatch==8.2
    # via ansible-lint
webencodings==0.5.1
    # via bleach
whitenoise==5.2.0
    # via pulpcore
xlrd==2.0.1
    # via tablib
xlwt==1.3.0
    # via tablib
yarl==1.6.3
    # via
    #   aiohttp
    #   pulpcore
y^C
[root@ip-10-0-15-31 pulp]# pwd
/usr/local/lib/pulp
[root@ip-10-0-15-31 pulp]# cat requirements.in 
pulpcore==3.14.2
pulp-container
pulp-ansible
git+https://github.com/ansible/galaxy_ng@master#egg=galaxy-ng

# Any plugins listed below were already installed but not specified in
# pulp_install_plugins
[root@ip-10-0-15-31 pulp]# bin/pip-compile 
#
# This file is autogenerated by pip-compile with python 3.8
# To update, run:
#
#    pip-compile
#
aiodns==3.0.0
    # via pulpcore
aiofiles==0.7.0
    # via pulpcore
aiohttp==3.7.4.post0
    # via pulpcore
aioredis==2.0.0
    # via pulpcore
ansible-core==2.11.3
    # via galaxy-importer
ansible-lint==5.1.2
    # via galaxy-importer
async-lru==1.0.2
    # via pulp-ansible
async-timeout==3.0.1
    # via
    #   aiohttp
    #   aioredis
asyncio-throttle==1.0.2
    # via pulpcore
attrs==21.2.0
    # via
    #   aiohttp
    #   galaxy-importer
    #   jsonschema
backoff==1.10.0
    # via pulpcore
bleach==3.3.1
    # via galaxy-importer
bleach-allowlist==1.0.3
    # via galaxy-importer
bracex==2.1.1
    # via wcmatch
certifi==2021.5.30
    # via requests
cffi==1.14.6
    # via
    #   cryptography
    #   pycares
chardet==4.0.0
    # via aiohttp
charset-normalizer==2.0.4
    # via requests
click==8.0.1
    # via
    #   pulpcore
    #   rq
colorama==0.4.4
    # via rich
commonmark==0.9.1
    # via rich
cryptography==3.4.7
    # via
    #   ansible-core
    #   pyjwt
defusedxml==0.7.1
    # via odfpy
diff-match-patch==20200713
    # via django-import-export
django==2.2.24
    # via
    #   django-currentuser
    #   django-filter
    #   django-guardian
    #   django-guid
    #   django-import-export
    #   django-lifecycle
    #   djangorestframework
    #   drf-nested-routers
    #   drf-spectacular
    #   galaxy-ng
    #   pulpcore
django-currentuser==0.5.3
    # via pulpcore
django-filter==2.4.0
    # via pulpcore
django-guardian==2.4.0
    # via pulpcore
django-guid==2.2.1
    # via pulpcore
django-import-export==2.5.0
    # via pulpcore
django-lifecycle==0.9.1
    # via pulpcore
django-prometheus==2.1.0
    # via galaxy-ng
djangorestframework==3.12.4
    # via
    #   drf-nested-routers
    #   drf-spectacular
    #   pulpcore
djangorestframework-queryfields==1.0.0
    # via pulpcore
drf-access-policy==0.9.2
    # via pulpcore
drf-nested-routers==0.93.3
    # via pulpcore
drf-spectacular==0.17.2
    # via
    #   galaxy-ng
    #   pulpcore
dynaconf==3.1.4
    # via pulpcore
ecdsa==0.13.3
    # via pulp-container
enrich==1.2.6
    # via ansible-lint
et-xmlfile==1.1.0
    # via openpyxl
flake8==3.9.2
    # via galaxy-importer
future==0.18.2
    # via pyjwkest
galaxy-importer==0.3.4
    # via
    #   galaxy-ng
    #   pulp-ansible
git+https://github.com/ansible/galaxy_ng@master#egg=galaxy-ng
    # via -r requirements.in
gunicorn==20.1.0
    # via pulpcore
idna==3.2
    # via
    #   requests
    #   yarl
inflection==0.5.1
    # via drf-spectacular
jinja2==3.0.1
    # via
    #   ansible-core
    #   pulpcore
jsonschema==3.2.0
    # via
    #   drf-spectacular
    #   pulp-ansible
markdown==3.3.4
    # via galaxy-importer
markuppy==1.14
    # via tablib
markupsafe==2.0.1
    # via jinja2
mccabe==0.6.1
    # via flake8
multidict==5.1.0
    # via
    #   aiohttp
    #   yarl
odfpy==1.4.1
    # via tablib
openpyxl==3.0.7
    # via tablib
packaging==21.0
    # via
    #   ansible-core
    #   ansible-lint
    #   bleach
    #   pulp-ansible
prometheus-client==0.11.0
    # via django-prometheus
psycopg2==2.8.6
    # via pulpcore
pulp-ansible==0.9.0
    # via
    #   -r requirements.in
    #   galaxy-ng
pulp-container==2.7.1
    # via
    #   -r requirements.in
    #   galaxy-ng
pulpcore==3.14.2
    # via
    #   -r requirements.in
    #   galaxy-ng
    #   pulp-ansible
    #   pulp-container
pycares==4.0.0
    # via aiodns
pycodestyle==2.7.0
    # via flake8
pycparser==2.20
    # via cffi
pycryptodomex==3.10.1
    # via pyjwkest
pyflakes==2.3.1
    # via flake8
pygments==2.9.0
    # via rich
pygtrie==2.4.2
    # via pulpcore
pyjwkest==1.4.2
    # via pulp-container
pyjwt[crypto]==1.7.1
    # via pulp-container
pyparsing==2.4.7
    # via
    #   drf-access-policy
    #   packaging
pyrsistent==0.18.0
    # via jsonschema
python-gnupg==0.4.7
    # via pulpcore
pytz==2021.1
    # via django
pyyaml==5.4.1
    # via
    #   ansible-core
    #   ansible-lint
    #   drf-spectacular
    #   galaxy-importer
    #   pulp-ansible
    #   pulpcore
    #   tablib
redis==3.5.3
    # via
    #   pulpcore
    #   rq
requests==2.26.0
    # via
    #   galaxy-importer
    #   pyjwkest
resolvelib==0.5.4
    # via ansible-core
rich==10.6.0
    # via
    #   ansible-lint
    #   enrich
rq==1.8.1
    # via pulpcore
ruamel.yaml==0.17.10
    # via ansible-lint
ruamel.yaml.clib==0.2.6
    # via ruamel.yaml
semantic-version==2.8.5
    # via
    #   galaxy-importer
    #   pulp-ansible
six==1.16.0
    # via
    #   bleach
    #   jsonschema
    #   pyjwkest
    #   url-normalize
sqlparse==0.4.1
    # via django
tablib[html,ods,xls,xlsx,yaml]==3.0.0
    # via django-import-export
tenacity==8.0.1
    # via ansible-lint
typing-extensions==3.10.0.0
    # via
    #   aiohttp
    #   aioredis
uritemplate==3.0.1
    # via drf-spectacular
url-normalize==1.4.3
    # via pulp-container
urllib3==1.26.6
    # via requests
urlman==2.0.1
    # via django-lifecycle
wcmatch==8.2
    # via ansible-lint
webencodings==0.5.1
    # via bleach
whitenoise==5.2.0
    # via pulpcore
xlrd==2.0.1
    # via tablib
xlwt==1.3.0
    # via tablib
yarl==1.6.3
    # via
    #   aiohttp
    #   pulpcore

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```
# The following packages are considered to be unsafe in a requirements file:
# setuptools